x-pack/platform/packages/shared/kbn-streams-schema/src/models/ingest/fields/index.ts

```diff
@@ -23,10 +23,14 @@ export const FIELD_DEFINITION_TYPES = [
 export type FieldDefinitionType = (typeof FIELD_DEFINITION_TYPES)[number];
 
 // We redefine "first class" parameters
-export type FieldDefinitionConfig = MappingProperty & {
-  type: FieldDefinitionType;
-  format?: string;
-};
+export type FieldDefinitionConfig =
+  | (MappingProperty & {
+      type: FieldDefinitionType;
+      format?: string;
+    })
+  | {
+      type: 'system';
```

Comment:
[nit] Could we include `system` as a constant in ` x-pack/platform/plugins/shared/streams/common/constants.ts`? If we do, we could use it from the other parts of the code where you check `something === 'system'`

Reply:
What would we gain from this? Typescript already makes sure that it's used correctly, I'm not sure a constant adds a lot here.

Reply:
or I was thinking that we could even create sort of a type guard for this one
```
export const systemFieldSchema = z.object({ type: z.literal('system') });

export const fieldDefinitionConfigSchema: z.Schema<FieldDefinitionConfig> = z.intersection(
  recursiveRecord,
  z.union([
    z.object({
      type: z.enum(FIELD_DEFINITION_TYPES),
      format: z.optional(NonEmptyString),
    }),
    systemFieldSchema,
  ])
);

export const isSystemField = createIsNarrowSchema(
  fieldDefinitionConfigSchema,
  systemFieldSchema
);
```
And use this directly instead of checking for `type === 'system'` everytime. In any case not a blocker

Reply:
Ah, got it - that would make sense, but unfortunately it doesn't work well in practice. If you check the places where I had to check for `system`, it's always a different type because we unpack and repack fields in a couple different ways depending on where you look at them.

That makes it difficult to use a single schema for it, but since the check is very straightforward and type safety doesn't suffer, I'm not too worried.